### PR TITLE
Fix issue with multiple targets and remove smfmac tests from unsupported test targets

### PR DESCRIPTION
**Pull request by junliume**
#1371

### CODE REVIEW

**Comment by junliume** [comment]
@illsilin @jakpiase please help to check if this check if complete, build multiple targets and tests multiple targets for CK is indeed tricky

**Review by carlushuang — DISMISSED** [review]
LGTM

File: include/ck/utility/amd_smfmac.hpp [hunk 1]
```diff
@@ -16,8 +20,15 @@ struct intrin_smfmac_f32_16x16x32f16<16, 16>
     __device__ static void
     Run(const half4_t& reg_a, const half8_t& reg_b, const int32_t& reg_idx, FloatC& reg_c)
     {
+#if defined(__gfx94__)
         reg_c.template AsType<float4_t>()(Number<0>{}) = __builtin_amdgcn_smfmac_f32_16x16x32_f16(
             reg_a, reg_b, reg_c.template AsType<float4_t>()[Number<0>{}], reg_idx, 0, 0);
+#else
+        ignore = reg_a;
```
Comment (junliume) [inline include/ck/utility/amd_smfmac.hpp:23]: These changes are consolidated from #1358

File: test/grouped_convnd_bwd_data/CMakeLists.txt [hunk 2]
```diff
@@ -2,11 +2,11 @@ add_gtest_executable(test_grouped_convnd_bwd_data test_grouped_convnd_bwd_data_x
 if(result EQUAL 0)
     target_link_libraries(test_grouped_convnd_bwd_data PRIVATE utility device_grouped_conv2d_bwd_data_instance device_grouped_conv3d_bwd_data_instance)
 endif()
-add_gtest_executable(test_grouped_convnd_bwd_data_interface test_grouped_convnd_bwd_data_interface_xdl.cpp)
```
Comment (junliume) [inline test/grouped_convnd_bwd_data/CMakeLists.txt:5]: For reviewers: these are fixing duplicate test name issue

File: include/ck/utility/amd_smfmac.hpp [hunk 3]
```diff
@@ -5,6 +5,10 @@
 #pragma once
 
 namespace ck {
+// Define the common macro for gfx94x models
+#if defined(__gfx940__) || defined(__gfx941__) || defined(__gfx942__)
+#define __gfx94__
+#endif
```
Comment (aosewski) [inline include/ck/utility/amd_smfmac.hpp]: Why we need this macro here? It's duplicated from https://github.com/ROCm/composable_kernel/blob/497ccb872b6a9a921c01df4dca49dac7cb242c72/include/ck/ck.hpp#L59-L60 which is included at the beginning of this file.

**Review by jakpiase — COMMENTED** [review]
From what I remember smfmac instructions won't work on earlier ROCm versions than 6.2. Is there any way to stop using this test on cmake level based on that condition?

**Comment by junliume** [comment]
> From what I remember smfmac instructions won't work on earlier ROCm versions than 6.2. Is there any way to stop using this test on cmake level based on that condition?

maybe `if(CK_hip_VERSION VERSION_GREATER_EQUAL 6.2)` ?

**Comment by junliume** [comment]
@jakpiase @aosewski please feel free to edit this branch in this PR :) I will be offline for the next a few hours.

All the remaining problems are easily reproducible by:
> CXX=/opt/rocm/bin/amdclang++ cmake -DCMAKE_PREFIX_PATH=/opt/rocm -DCMAKE_BUILD_TYPE=Release -DGPU_TARGETS="gfx1100;gfx90a;gfx942" ..

and then
> make -j$(nproc) check

File: client_example/25_wrapper/CMakeLists.txt [hunk 4]
```diff
@@ -2,7 +2,7 @@ add_executable(client_tensor_transform_using_wrapper tensor_transform_using_wrap
 target_link_libraries(client_tensor_transform_using_wrapper PRIVATE composable_kernel::device_other_operations)
 add_executable(client_wrapper_img2col wrapper_img2col.cpp)
 target_link_libraries(client_wrapper_img2col PRIVATE composable_kernel::device_other_operations)
-if(GPU_TARGETS MATCHES "gfx9")
```
Comment (junliume) [inline client_example/25_wrapper/CMakeLists.txt]: FYI reviewers: this is not ideal :(

described here: https://github.com/ROCm/composable_kernel/issues/1371#issuecomment-2207020521
When we have a mixed multi-targets such as `-DGPU_TARGETS="gfx1100;gfx90a"`:
- Compiler will iterate over all named targets and try to build the client example 25. However, too many places directly uses XDL even inline, so this example should not be built with unsupported platforms;
- However, **if the goal is to build once, and run everywhere**:
    - **[Against what this change is doing]** we should still build this client-example, and at runtime check applicability and decide to skip or not;
    - However, IDK how to control compile time iterations, and skip certain lines/files based on which micro-architecture is been compiled for. e.g. `-DGPU_TARGETS="gfx1100;gfx90a"` means that compiler will compile for `--offload-arch=gfx1100 --offload-arch=gfx90a` in two iterations (and combine results together afterward). **is there any control to skip lines/files by MACRO or something similar in one of these compilation iterations?**

Comment (junliume) [reply client_example/25_wrapper/CMakeLists.txt]: This is attempting to dispatch compile time multi-targets properly:
https://github.com/ROCm/composable_kernel/pull/1372/commits/8ad46ed361b28362b59b13ecc4bd091fae9e7107

File: client_example/25_wrapper/wrapper_basic_gemm.cpp [hunk 5]
```diff
@@ -1,5 +1,7 @@
 // SPDX-License-Identifier: MIT
 // Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
+#if(!defined(__HIP_DEVICE_COMPILE__) || defined(__gfx908__) || defined(__gfx90a__) || \
+    defined(__gfx94__))
```
Comment (atamazov) [inline client_example/25_wrapper/wrapper_basic_gemm.cpp]: @junliume The expected effect is to compile the code for the **_host_** and for **_MI100/200/300_**, right? Looks good (at least for clang 19, see https://clang.llvm.org/docs/AMDGPUSupport.html) except `__gfx94__` which looks suspicious. If you are in doubt too, then you can try this:

```C++
#if(!defined(__HIP_DEVICE_COMPILE__) || (defined(__gfx9__) && !defined(__gfx906__) && !defined(__gfx900__))
```
The above should enable compilation for all gfx9 parts except Vega10/20.

BTW it's a shame that compiler doesn't have the `__HIP_ARCH_*` feature macro that indicates the presence of xDLOPs (MAI) on target, see https://rocm.docs.amd.com/projects/HIP/en/latest/how-to/hip_porting_guide.html#table-of-architecture-properties. Perhaps it's worth requesting it.

Comment (atamazov) [reply client_example/25_wrapper/wrapper_basic_gemm.cpp]: @junliume Maybe it is worth introducing intermediate macros that would enable cleaner programming, e.g. put something like this into a header (target_features.hpp?):
```C++
#define TARGET_HOST   1
#define TARGET_GFX9   0
#define TARGET_GFX900 0
#define TARGET_GFX906 0
#define TARGET_GFX908 0
#define TARGET_GFX90A 0
#define TARGET_GFX941 0
#define TARGET_GFX942 0
...
#if defined(__HIP_DEVICE_COMPILE__)
  #undef TARGET_HOST
  #define TARGET_HOST 0
  #if defined(__gfx9__)
    #undef TARGET_GFX9
    #define TARGET_GFX9 1
  #endif
  ...
  #if defined(__gfx906__)
    #undef TARGET_GFX906
    #define TARGET_GFX906 1
  #endif
  #if defined(__gfx908__)
    #undef TARGET_GFX908
    #define TARGET_GFX908 1
  #endif
  ...
  #if defined(__gfx940__)
    #undef TARGET_GFX940
    #define TARGET_GFX940 1
  #endif
  #if defined(__gfx941__)
    #undef TARGET_GFX941
    #define TARGET_GFX941 1
  #endif
  ...
#endif

// Now define features:
#define TARGET_GFX94X (TARGET_GFX941 || TARGET_GFX942 || /* enumerate all gfx94s */ )
#define TARGET_GFX_XDLOPS (TARGET_GFX908 || TARGET_GFX90A || TARGET_GFX94X /* add others in the future*/ )
```

Then you can use this like:
```C++
#include <target_features.hpp>

#if TARGET_HOST || (TARGET_XDLOPS && TARGET_GFX9)
  // Enabled for host and for GFX9 XDLOPS devices...
```
which seems much cleaner to me.

Comment (junliume) [reply client_example/25_wrapper/wrapper_basic_gemm.cpp]: @atamazov found these lines so maybe we should use existing ones there
https://github.com/ROCm/composable_kernel/blob/fix_1371/include/ck/ck.hpp#L55-L74

Comment (atamazov) [reply client_example/25_wrapper/wrapper_basic_gemm.cpp]: @junliume At first glance, yes. And you may even find some suitable "feature" macros there, like `CK_USE_AMD_MFMA_BF16_1K_OP` or so (but I am still unable to find things like TARGET_GFX_XDLOPS there)))

However, these just _define_, which enforces to use `defined()` everywhere. What I am proposing is _always defining_ to either 0 or 1, which allows for cleaner constructs. Just to mention.

Comment (junliume) [reply client_example/25_wrapper/wrapper_basic_gemm.cpp]: > @junliume At first glance, yes. And you may even find some suitable "feature" macros there, like `CK_USE_AMD_MFMA_BF16_1K_OP` or so (but I am still unable to find things like TARGET_GFX_XDLOPS there)))
> 
> However, these just _define_, which enforces to use `defined()` everywhere. What I am proposing is _always defining_ to either 0 or 1, which allows for cleaner constructs. Just to mention.

Good point, I will keep a note about this refactoring and assign CK developer on it.

**Review by carlushuang — APPROVED** [review]
LGTM

**Pull request merged**